src/main/java/com/autotune/analyzer/plots/PlotManager.java

```diff
@@ -58,37 +57,46 @@ public PlotData.PlotsData generatePlots() {
             calendar.add(Calendar.MILLISECOND, (int) millisecondsToAdd);
             // Convert the modified Calendar back to a Timestamp
             Timestamp newTimestamp = new Timestamp(calendar.getTimeInMillis());
-            PlotData.UsageData cpuUsage = getUsageData(sortedResultsHashMap.subMap(newTimestamp, true, incrementStartTime, true), AnalyzerConstants.MetricName.cpuUsage, "cores");
-            PlotData.UsageData memoryUsage = getUsageData(sortedResultsHashMap.subMap(newTimestamp, true, incrementStartTime, true), AnalyzerConstants.MetricName.memoryUsage, "MiB");
+            PlotData.UsageData cpuUsage = getUsageData(sortedResultsHashMap.subMap(newTimestamp, true,
+                    incrementStartTime,false), AnalyzerConstants.MetricName.cpuUsage);
+            PlotData.UsageData memoryUsage = getUsageData(sortedResultsHashMap.subMap(newTimestamp, true,
+                    incrementStartTime, false), AnalyzerConstants.MetricName.memoryUsage);
             plotsDataMap.put(newTimestamp, new PlotData.PlotPoint(cpuUsage, memoryUsage));
             incrementStartTime = newTimestamp;
         }
 
         return new PlotData.PlotsData(recommendationTerm.getPlots_datapoints(), plotsDataMap);
     }
 
-    PlotData.UsageData getUsageData(Map<Timestamp, IntervalResults> resultInRange, AnalyzerConstants.MetricName metricName, String format) {
-        // Extract CPU values
-        List<Double> cpuValues = resultInRange.values().stream()
-                .filter(intervalResults -> intervalResults.getMetricResultsMap().containsKey(metricName))
-                .mapToDouble(intervalResults -> {
-                    MetricResults metricResults = intervalResults.getMetricResultsMap().get(metricName);
-                    return (metricResults != null && metricResults.getAggregationInfoResult() != null) ? metricResults.getAggregationInfoResult().getSum() : 0.0;
-                })
-                .boxed() // Convert double to Double
-                .collect(Collectors.toList());
-        if (cpuValues.size() > 0) {
-            double q1 = CommonUtils.percentile(TWENTYFIVE_PERCENTILE, cpuValues);
-            double q3 = CommonUtils.percentile(SEVENTYFIVE_PERCENTILE, cpuValues);
-            double median = CommonUtils.percentile(FIFTY_PERCENTILE, cpuValues);
+    PlotData.UsageData getUsageData(Map<Timestamp, IntervalResults> resultInRange, AnalyzerConstants.MetricName metricName) {
+        // stream through the results value and extract the CPU values
+        if (metricName.equals(AnalyzerConstants.MetricName.cpuUsage)) {
+            List<Double> cpuValues = CostBasedRecommendationModel.getCPUUsageList(resultInRange);
+            return getPercentileData(cpuValues, resultInRange, metricName);
+        } else {
+            // stream through the results value and extract the memory values
+            List<Double> memUsageList = resultInRange.values()
```

Comment:
instead of stream can we do like this  ? 
List<Double> memUsageList = ostBasedRecommendationModel.calculateMemoryUsage(resultInRange)

Reply:
Code is refactored now.